### PR TITLE
android: get java env from thread local storage

### PR DESCRIPTION
**Pull request by weihuoya**

### CODE REVIEW

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

File: Source/Android/jni/AndroidCommon/IDCache.cpp [hunk 1]
```diff
@@ -29,9 +29,27 @@ static jmethodID s_do_rumble;
...
+
+    ~OwnedEnv()
+    {
+      if (status == JNI_EDETACHED)
+        s_java_vm->DetachCurrentThread();
+    }
+
+    int status;
+    JNIEnv* env;
+  } ownedEnv;
```
Comment (JosJuice) [inline Source/Android/jni/AndroidCommon/IDCache.cpp]: Please rename `ownedEnv` to `owned_env`.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
Rebase on the latest version of master in order to fix the "AfterCaseLabel" lint problem.

File: Source/Android/jni/AndroidCommon/IDCache.cpp [hunk 2]
```diff
@@ -29,9 +29,27 @@ static jmethodID s_do_rumble;
 
 namespace IDCache
 {
-JavaVM* GetJavaVM()
-{
-  return s_java_vm;
+JNIEnv* GetEnvForThread()
+{
+  thread_local static struct owned_env
```
Comment (JosJuice) [inline Source/Android/jni/AndroidCommon/IDCache.cpp]: Sorry, maybe I wasn't clear with my earlier comment. The struct name should still be `OwnedEnv`. The underscore style of naming is only used for variable names.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

File: Source/Android/jni/AndroidCommon/IDCache.cpp [hunk 3]
```diff
@@ -29,9 +29,27 @@ static jmethodID s_do_rumble;
...
 namespace IDCache
 {
-JavaVM* GetJavaVM()
-{
-  return s_java_vm;
+JNIEnv* GetEnvForThread()
+{
+  thread_local static struct OwnedEnv
+  {
+    OwnedEnv() : env(nullptr)
```
Comment (lioncash) [inline Source/Android/jni/AndroidCommon/IDCache.cpp]: This can be joined at the declaration like:

```cpp
JNIEnv* env = nullptr;
```

**Comment by stenzek** [comment]
@dolphin-emu-bot rebuild

**Comment by farouk327** [comment]
The emulator stop launching games now and the same for the future builds

**Comment by JosJuice** [comment]
And you've bisected it down to this change? What device is this happening on?

**Comment by farouk327** [comment]
Lenovo k5 pro
Sorry for that i am new in github

**Comment by farouk327** [comment]
That Chinese son of the bitch fucked up the emulator he is no more function-able

**Comment by farouk327** [comment]
Fix the fucking emulator you damned Chinese i hate you weihuoya 😠die with corona s.o.b

**Pull request merged**